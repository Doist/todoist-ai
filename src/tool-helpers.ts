import {
    type MoveTaskArgs,
    type PersonalProject,
    type Task,
    type TodoistApi,
    type WorkspaceProject,
    getSanitizedContent,
} from '@doist/todoist-api-typescript'
import z from 'zod'
import { formatDuration } from './utils/duration-parser.js'

export type Project = PersonalProject | WorkspaceProject

export function isPersonalProject(project: Project): project is PersonalProject {
    return 'inboxProject' in project
}

export function isWorkspaceProject(project: Project): project is WorkspaceProject {
    return 'accessLevel' in project
}

/**
 * Creates a MoveTaskArgs object from move parameters, validating that exactly one is provided.
 * @param taskId - The task ID (used for error messages)
 * @param projectId - Optional project ID to move to
 * @param sectionId - Optional section ID to move to
 * @param parentId - Optional parent ID to move to
 * @returns MoveTaskArgs object with exactly one destination
 * @throws Error if multiple move parameters are provided or none are provided
 */
export function createMoveTaskArgs(
    taskId: string,
    projectId?: string,
    sectionId?: string,
    parentId?: string,
): MoveTaskArgs {
    // Validate that only one move parameter is provided (RequireExactlyOne constraint)
    const moveParams = [projectId, sectionId, parentId].filter(Boolean)
    if (moveParams.length > 1) {
        throw new Error(
            `Task ${taskId}: Only one of projectId, sectionId, or parentId can be specified at a time. The Todoist API requires exactly one destination for move operations.`,
        )
    }

    if (moveParams.length === 0) {
        throw new Error(
            `Task ${taskId}: At least one of projectId, sectionId, or parentId must be provided for move operations.`,
        )
    }

    // Build moveArgs with the single defined value
    if (projectId) return { projectId }
    if (sectionId) return { sectionId }
    if (parentId) return { parentId }

    // This should never be reached due to the validation above
    throw new Error('Unexpected error: No valid move parameter found')
}

/**
 * Map a single Todoist task to a more structured format, for LLM consumption.
 * @param task - The task to map.
 * @returns The mapped task.
 */
function mapTask(task: Task) {
    return {
        id: task.id,
        content: getSanitizedContent(task.content),
        description: getSanitizedContent(task.description),
        dueDate: task.due?.date,
        recurring: task.due?.isRecurring && task.due.string ? task.due.string : false,
        priority: task.priority,
        projectId: task.projectId,
        sectionId: task.sectionId,
        parentId: task.parentId,
        labels: task.labels,
        duration: task.duration ? formatDuration(task.duration.amount) : null,
    }
}

/**
 * Map a single Todoist project to a more structured format, for LLM consumption.
 * @param project - The project to map.
 * @returns The mapped project.
 */
function mapProject(project: Project) {
    return {
        id: project.id,
        name: project.name,
        color: project.color,
        isFavorite: project.isFavorite,
        isShared: project.isShared,
        parentId: isPersonalProject(project) ? (project.parentId ?? null) : null,
        inboxProject: isPersonalProject(project) ? (project.inboxProject ?? false) : false,
        viewStyle: project.viewStyle,
    }
}

const ErrorSchema = z.object({
    httpStatusCode: z.number(),
    responseData: z.object({
        error: z.string(),
        errorCode: z.number(),
        errorTag: z.string(),
    }),
})

async function getTasksByFilter({
    client,
    query,
    limit,
    cursor,
}: {
    client: TodoistApi
    query: string
    limit: number | undefined
    cursor: string | undefined
}) {
    try {
        const { results, nextCursor } = await client.getTasksByFilter({
            query,
            cursor,
            limit,
        })

        return {
            tasks: results.map(mapTask),
            nextCursor,
        }
    } catch (error) {
        const parsedError = ErrorSchema.safeParse(error)
        if (!parsedError.success) {
            throw error
        }
        const { responseData } = parsedError.data
        if (responseData.errorTag === 'INVALID_SEARCH_QUERY') {
            throw new Error(`Invalid filter query: ${query}`)
        }
        throw new Error(
            `${responseData.error} (tag: ${responseData.errorTag}, code: ${responseData.errorCode})`,
        )
    }
}

export { getTasksByFilter, mapTask, mapProject }
