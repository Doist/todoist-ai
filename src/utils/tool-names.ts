/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */

export const TOOL_NAMES = {
    // Task management tools
    TASKS_ADD_MULTIPLE: 'tasks-add-multiple',
    TASKS_COMPLETE_MULTIPLE: 'tasks-complete-multiple',
    TASKS_UPDATE_MULTIPLE: 'tasks-update-multiple',
    TASKS_SEARCH: 'tasks-search',
    TASKS_LIST_BY_DATE: 'tasks-list-by-date',
    TASKS_LIST_COMPLETED: 'tasks-list-completed',
    TASKS_LIST_FOR_CONTAINER: 'tasks-list-for-container',

    // Project management tools
    PROJECTS_LIST: 'projects-list',
    PROJECTS_MANAGE: 'projects-manage',

    // Section management tools
    SECTIONS_SEARCH: 'sections-search',
    SECTIONS_MANAGE: 'sections-manage',

    // General tools
    OVERVIEW: 'overview',
    DELETE_ONE: 'delete-one',
} as const

// Type for all tool names
export type ToolName = (typeof TOOL_NAMES)[keyof typeof TOOL_NAMES]
