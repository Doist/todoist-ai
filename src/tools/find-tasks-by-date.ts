import { addDays, formatISO } from 'date-fns'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { getTasksByFilter } from '../tool-helpers.js'
import { ApiLimits } from '../utils/constants.js'
import { LabelsSchema, generateLabelsFilter } from '../utils/labels.js'
import {
    generateTaskNextSteps,
    getDateString,
    previewTasks,
    summarizeList,
} from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const ArgsSchema = {
    startDate: z
        .string()
        .regex(/^(\d{4}-\d{2}-\d{2}|today|overdue)$/)
        .describe(
            "The start date to get the tasks for. Format: YYYY-MM-DD, 'today', or 'overdue'.",
        ),
    daysCount: z
        .number()
        .int()
        .min(1)
        .max(30)
        .default(1)
        .describe(
            "The number of days to get the tasks for, starting from the start date. Ignored when startDate is 'overdue'.",
        ),
    limit: z
        .number()
        .int()
        .min(1)
        .max(ApiLimits.TASKS_MAX)
        .default(ApiLimits.TASKS_DEFAULT)
        .describe('The maximum number of tasks to return.'),
    cursor: z
        .string()
        .optional()
        .describe(
            'The cursor to get the next page of tasks (cursor is obtained from the previous call to this tool, with the same parameters).',
        ),
    ...LabelsSchema,
}

const findTasksByDate = {
    name: ToolNames.FIND_TASKS_BY_DATE,
    description:
        "Get tasks by date range or overdue tasks. Use startDate 'overdue' for overdue tasks, or provide a date/date range.",
    parameters: ArgsSchema,
    async execute(args, client) {
        let query = ''

        if (args.startDate === 'overdue') {
            query = 'overdue'
        } else {
            const startDate =
                args.startDate === 'today'
                    ? formatISO(new Date(), { representation: 'date' })
                    : args.startDate
            const endDate = addDays(startDate, args.daysCount + 1)
            const endDateStr = formatISO(endDate, { representation: 'date' })
            query = `(due after: ${startDate} | due: ${startDate}) & due before: ${endDateStr}`
        }

        const labelsFilter = generateLabelsFilter(args.labels, args.labelsOperator)
        if (labelsFilter.length > 0) {
            // If there is already a query, we need to append the & operator first
            if (query.length > 0) query += ' & '
            // Add the labels to the filter
            query += `(${labelsFilter})`
        }

        const result = await getTasksByFilter({
            client,
            query,
            cursor: args.cursor,
            limit: args.limit,
        })

        const textContent = generateTextContent({
            tasks: result.tasks,
            args,
            nextCursor: result.nextCursor,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                tasks: result.tasks,
                nextCursor: result.nextCursor,
                totalCount: result.tasks.length,
                hasMore: Boolean(result.nextCursor),
                appliedFilters: args,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    tasks,
    args,
    nextCursor,
}: {
    tasks: Awaited<ReturnType<typeof getTasksByFilter>>['tasks']
    args: z.infer<z.ZodObject<typeof ArgsSchema>>
    nextCursor: string | null
}) {
    // Generate filter description
    const filterHints: string[] = []
    if (args.startDate === 'overdue') {
        filterHints.push('overdue tasks only')
    } else if (args.startDate === 'today') {
        filterHints.push(`today${args.daysCount > 1 ? ` + ${args.daysCount - 1} more days` : ''}`)
    } else {
        filterHints.push(
            `${args.startDate}${args.daysCount > 1 ? ` to ${getDateString(addDays(args.startDate, args.daysCount))}` : ''}`,
        )
    }

    // Add label filter information
    if (args.labels && args.labels.length > 0) {
        const labelText = args.labels
            .map((label) => `@${label}`)
            .join(args.labelsOperator === 'and' ? ' & ' : ' | ')
        filterHints.push(`labels: ${labelText}`)
    }

    // Generate subject description
    const subject =
        args.startDate === 'overdue'
            ? 'Overdue tasks'
            : args.startDate === 'today'
              ? `Today's tasks`
              : `Tasks for ${args.startDate}`

    // Generate helpful suggestions for empty results
    const zeroReasonHints: string[] = []
    if (tasks.length === 0) {
        if (args.startDate === 'overdue') {
            zeroReasonHints.push('Great job! No overdue tasks')
            zeroReasonHints.push("Check today's tasks with startDate='today'")
        } else {
            zeroReasonHints.push("Expand date range with larger 'daysCount'")
            zeroReasonHints.push("Check 'overdue' for past-due items")
        }
    }

    // Generate contextual next steps
    const now = new Date()
    const todayStr = getDateString(now)
    const nextSteps = generateTaskNextSteps('listed', tasks, {
        hasToday: args.startDate === 'today' || tasks.some((task) => task.dueDate === todayStr),
        hasOverdue:
            args.startDate === 'overdue' ||
            tasks.some((task) => task.dueDate && new Date(task.dueDate) < now),
    })

    return summarizeList({
        subject,
        count: tasks.length,
        limit: args.limit,
        nextCursor: nextCursor ?? undefined,
        filterHints,
        previewLines: previewTasks(tasks),
        zeroReasonHints,
        nextSteps,
    })
}

export { findTasksByDate }
