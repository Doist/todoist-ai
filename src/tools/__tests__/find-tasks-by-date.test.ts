import type { TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { getTasksByFilter } from '../../tool-helpers.js'
import {
    TEST_ERRORS,
    TEST_IDS,
    createMappedTask,
    extractStructuredContent,
    extractTextContent,
} from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { findTasksByDate } from '../find-tasks-by-date.js'

// Mock the tool helpers
jest.mock('../../tool-helpers', () => ({
    getTasksByFilter: jest.fn(),
}))

const mockGetTasksByFilter = getTasksByFilter as jest.MockedFunction<typeof getTasksByFilter>

// Mock the Todoist API (not directly used by find-tasks-by-date, but needed for type)
const mockTodoistApi = {} as TodoistApi

// Mock date-fns functions to make tests deterministic
jest.mock('date-fns', () => ({
    addDays: jest.fn(() => new Date('2025-08-16')), // Return predictable end date
    formatISO: jest.fn((date, options) => {
        if (typeof date === 'string') {
            return date // Return string dates as-is
        }
        if (
            options &&
            typeof options === 'object' &&
            'representation' in options &&
            options.representation === 'date'
        ) {
            return '2025-08-15' // Return predictable date for 'today'
        }
        return '2025-08-16' // Return predictable end date
    }),
}))

const { FIND_TASKS_BY_DATE, UPDATE_TASKS } = ToolNames

describe(`${FIND_TASKS_BY_DATE} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
        // Mock current date to make tests deterministic
        jest.spyOn(Date, 'now').mockReturnValue(new Date('2025-08-15T10:00:00Z').getTime())
    })

    afterEach(() => {
        jest.restoreAllMocks()
    })

    describe('listing overdue tasks', () => {
        it.each([
            { daysCount: 7, hasTasks: true, description: 'with tasks' },
            { daysCount: 5, hasTasks: false, description: 'ignoring daysCount' },
        ])('should handle overdue tasks $description', async ({ daysCount, hasTasks }) => {
            const mockTasks = hasTasks
                ? [
                      createMappedTask({
                          id: TEST_IDS.TASK_1,
                          content: 'Overdue task',
                          dueDate: '2025-08-10',
                          priority: 2,
                          labels: ['urgent'],
                      }),
                  ]
                : []

            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(
                { startDate: 'overdue', limit: 50, daysCount, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: 'overdue',
                cursor: undefined,
                limit: 50,
            })
            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent.tasks).toHaveLength(hasTasks ? 1 : 0)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    totalCount: hasTasks ? 1 : 0,
                    hasMore: false,
                    nextCursor: null,
                    appliedFilters: expect.objectContaining({
                        startDate: 'overdue',
                    }),
                }),
            )
        })
    })

    describe('listing tasks by date range', () => {
        it('should get tasks for today when startDate is "today"', async () => {
            const mockTasks = [createMappedTask({ content: 'Today task', dueDate: '2025-08-15' })]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(
                { startDate: 'today', limit: 50, daysCount: 7, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query:
                    expect.stringContaining('due after:') && expect.stringContaining('due before:'),
                cursor: undefined,
                limit: 50,
            })
            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it.each([
            {
                name: 'specific date',
                params: { startDate: '2025-08-20', limit: 50, daysCount: 7 },
                tasks: [createMappedTask({ content: 'Specific date task', dueDate: '2025-08-20' })],
                cursor: null,
            },
            {
                name: 'multiple days with pagination',
                params: {
                    startDate: '2025-08-20',
                    daysCount: 3,
                    limit: 20,
                    cursor: 'current-cursor',
                },
                tasks: [
                    createMappedTask({
                        id: TEST_IDS.TASK_2,
                        content: 'Multi-day task 1',
                        dueDate: '2025-08-20',
                    }),
                    createMappedTask({
                        id: TEST_IDS.TASK_3,
                        content: 'Multi-day task 2',
                        dueDate: '2025-08-21',
                    }),
                ],
                cursor: 'next-page-cursor',
            },
        ])('should handle $name', async ({ params, tasks, cursor }) => {
            const mockResponse = { tasks, nextCursor: cursor }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(
                { ...params, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: expect.stringContaining('2025-08-20'),
                cursor: params.cursor || undefined,
                limit: params.limit,
            })
            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('pagination and limits', () => {
        it.each([
            {
                name: 'pagination parameters',
                params: {
                    startDate: 'today',
                    limit: 25,
                    daysCount: 7,
                    cursor: 'pagination-cursor',
                },
                expectedCursor: 'pagination-cursor',
                expectedLimit: 25,
            },
            {
                name: 'default values',
                params: { startDate: '2025-08-15', limit: 50, daysCount: 7 },
                expectedCursor: undefined,
                expectedLimit: 50,
            },
        ])('should handle $name', async ({ params, expectedCursor, expectedLimit }) => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            await findTasksByDate.execute(
                { ...params, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: expect.any(String),
                cursor: expectedCursor,
                limit: expectedLimit,
            })
        })
    })

    describe('edge cases', () => {
        it.each([
            { name: 'empty results', daysCount: 7, shouldReturnResult: true },
            { name: 'maximum daysCount', daysCount: 30, shouldReturnResult: false },
            { name: 'minimum daysCount', daysCount: 1, shouldReturnResult: false },
        ])('should handle $name', async ({ daysCount, shouldReturnResult }) => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const startDate = daysCount === 7 ? 'today' : '2025-08-15'
            const result = await findTasksByDate.execute(
                { startDate, limit: 50, daysCount, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledTimes(1)
            if (shouldReturnResult) {
                // Verify result is a concise summary
                expect(extractTextContent(result)).toMatchSnapshot()
            }
        })
    })

    describe('next steps logic', () => {
        it('should suggest appropriate actions when hasOverdue is true', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Overdue task from list',
                    dueDate: '2025-08-10', // Past date - creates hasOverdue context
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(
                {
                    startDate: '2025-08-15',
                    limit: 10,
                    daysCount: 1,
                    labels: [],
                    labelsOperator: 'or',
                },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(`Use ${UPDATE_TASKS} to modify priorities or due dates`)
        })

        it('should suggest today-focused actions when startDate is today', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: "Today's task",
                    dueDate: '2025-08-15', // Today's date based on our mock
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(
                { startDate: 'today', limit: 10, daysCount: 1, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(`Use ${UPDATE_TASKS} to modify priorities or due dates`)
        })

        it('should suggest appropriate actions when startDate is overdue', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Overdue task',
                    dueDate: '2025-08-10',
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(
                { startDate: 'overdue', limit: 10, daysCount: 1, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(`Use ${UPDATE_TASKS} to modify priorities or due dates`)
        })

        it('should provide helpful suggestions for empty overdue results', async () => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(
                { startDate: 'overdue', limit: 10, daysCount: 1, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Great job! No overdue tasks')
            expect(textContent).toContain("Check today's tasks with startDate='today'")
        })

        it('should provide helpful suggestions for empty date range results', async () => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(
                {
                    startDate: '2025-08-20',
                    limit: 10,
                    daysCount: 1,
                    labels: [],
                    labelsOperator: 'or',
                },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain("Expand date range with larger 'daysCount'")
            expect(textContent).toContain("Check 'overdue' for past-due items")
        })
    })

    describe('label filtering', () => {
        it.each([
            {
                name: 'single label with OR operator',
                params: {
                    startDate: 'today',
                    daysCount: 1,
                    limit: 50,
                    labels: ['@work'],
                    labelsOperator: 'or' as const,
                },
                expectedQueryPattern: '((@work))', // Will be combined with date query
            },
            {
                name: 'multiple labels with AND operator',
                params: {
                    startDate: 'overdue',
                    daysCount: 1,
                    limit: 50,
                    labels: ['@work', '@urgent'],
                    labelsOperator: 'and' as const,
                },
                expectedQueryPattern: 'overdue & ((@work  &  @urgent))',
            },
            {
                name: 'multiple labels with OR operator',
                params: {
                    startDate: '2025-08-20',
                    daysCount: 3,
                    limit: 50,
                    labels: ['@personal', '@shopping'],
                    labelsOperator: 'or' as const,
                },
                expectedQueryPattern: '((@personal  |  @shopping))',
            },
        ])('should filter tasks by labels: $name', async ({ params, expectedQueryPattern }) => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Task with work label',
                    labels: ['work'],
                    dueDate: '2025-08-20',
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(params, mockTodoistApi)

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: expect.stringContaining('(@'),
                cursor: undefined,
                limit: 50,
            })

            // For overdue specifically, check the exact pattern
            if (params.startDate === 'overdue') {
                expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                    client: mockTodoistApi,
                    query: expectedQueryPattern,
                    cursor: undefined,
                    limit: 50,
                })
            }

            const structuredContent = extractStructuredContent(result)
            expect(structuredContent.appliedFilters).toEqual(
                expect.objectContaining({
                    labels: params.labels,
                    labelsOperator: params.labelsOperator,
                }),
            )
        })

        it('should handle empty labels array', async () => {
            const params = {
                startDate: 'today' as const,
                daysCount: 1,
                limit: 50,
                labels: [],
                labelsOperator: 'or' as const,
            }

            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            await findTasksByDate.execute(params, mockTodoistApi)

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: expect.not.stringContaining('@'),
                cursor: undefined,
                limit: 50,
            })
        })

        it('should combine date filters with label filters', async () => {
            const params = {
                startDate: '2025-08-15' as const,
                daysCount: 1,
                limit: 25,
                labels: ['@important'],
                labelsOperator: 'or' as const,
            }

            const mockTasks = [
                createMappedTask({
                    content: 'Important task for specific date',
                    labels: ['important'],
                    dueDate: '2025-08-15',
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasksByDate.execute(params, mockTodoistApi)

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query:
                    expect.stringContaining('due after:') &&
                    expect.stringContaining('(@important)'),
                cursor: undefined,
                limit: 25,
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
        })
    })

    describe('error handling', () => {
        it.each([
            {
                error: TEST_ERRORS.INVALID_FILTER,
                params: { startDate: 'today', limit: 50, daysCount: 7 },
            },
            {
                error: TEST_ERRORS.API_RATE_LIMIT,
                params: { startDate: 'overdue', limit: 50, daysCount: 7 },
            },
            {
                error: TEST_ERRORS.INVALID_CURSOR,
                params: {
                    startDate: '2025-08-15',
                    limit: 50,
                    daysCount: 7,
                    cursor: 'invalid-cursor',
                },
            },
        ])('should propagate $error', async ({ error, params }) => {
            mockGetTasksByFilter.mockRejectedValue(new Error(error))
            await expect(
                findTasksByDate.execute(
                    { ...params, labels: [], labelsOperator: 'or' },
                    mockTodoistApi,
                ),
            ).rejects.toThrow(error)
        })
    })
})
