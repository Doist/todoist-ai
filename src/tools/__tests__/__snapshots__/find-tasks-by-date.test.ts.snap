// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`find-tasks-by-date tool edge cases should handle empty results 1`] = `
"Today's tasks + overdue: 0 (limit 50).
Filter: today + overdue tasks + 6 more days.
No results. Great job! No tasks for today or overdue."
`;

exports[`find-tasks-by-date tool label filtering should combine date filters with label filters 1`] = `
"Tasks for 2025-08-15: 1 (limit 25).
Filter: 2025-08-15; labels: @important.
Preview:
    Important task for specific date • due 2025-08-15 • P4 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`find-tasks-by-date tool listing tasks by date range should get tasks for today when startDate is "today" (includes overdue) 1`] = `
"Today's tasks + overdue: 1 (limit 50).
Filter: today + overdue tasks + 6 more days.
Preview:
    Today task • due 2025-08-15 • P4 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`find-tasks-by-date tool listing tasks by date range should handle multiple days with pagination 1`] = `
"Tasks for 2025-08-20: 2 (limit 20), more available.
Filter: 2025-08-20 to 2025-08-16.
Preview:
    Multi-day task 1 • due 2025-08-20 • P4 • id=8485093749
    Multi-day task 2 • due 2025-08-21 • P4 • id=8485093750
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Focus on overdue items first to get back on track
- Pass cursor 'next-page-cursor' to fetch more results."
`;

exports[`find-tasks-by-date tool listing tasks by date range should handle specific date 1`] = `
"Tasks for 2025-08-20: 1 (limit 50).
Filter: 2025-08-20 to 2025-08-16.
Preview:
    Specific date task • due 2025-08-20 • P4 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`find-tasks-by-date tool next steps logic should provide helpful suggestions for empty date range results 1`] = `
"Tasks for 2025-08-20: 0 (limit 10).
Filter: 2025-08-20.
No results. Expand date range with larger 'daysCount'; Check today's tasks with startDate='today'."
`;

exports[`find-tasks-by-date tool next steps logic should provide helpful suggestions for empty today results 1`] = `
"Today's tasks + overdue: 0 (limit 10).
Filter: today + overdue tasks.
No results. Great job! No tasks for today or overdue."
`;

exports[`find-tasks-by-date tool next steps logic should suggest appropriate actions when hasOverdue is true 1`] = `
"Tasks for 2025-08-15: 1 (limit 10).
Filter: 2025-08-15.
Preview:
    Overdue task from list • due 2025-08-10 • P4 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`find-tasks-by-date tool next steps logic should suggest today-focused actions when startDate is today 1`] = `
"Today's tasks + overdue: 1 (limit 10).
Filter: today + overdue tasks.
Preview:
    Today's task • due 2025-08-15 • P4 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Focus on overdue items first to get back on track"
`;
